### PR TITLE
feat: support persistent LIST properties for storage

### LINKED ISSUE #159: Introduce support for `List` type in storage.
Support List type in storage, whose component_type could also be a list.

### PR DESCRIPTION
**Pull request by zhanglei1949**
## What do these changes do?

Add end-to-end support for persistent `LIST<T>` properties (recursively `LIST` / `ARRAY` / storable-scalar children) on vertex and non-bundled edge schemas. Supersedes #157.

- **Storage**: new `ListPropertyColumn` with per-level `offsets + lengths + elements`. Same-length updates overwrite in place; length changes append; checkpoint dump compacts live spans only. Clone/Detach propagate to all three sub-columns for TP COW isolation.
- **Types / DDL**: column factory branch, YAML encode (including the previously missing TIMESTAMP/INTERVAL branches), common-PB and reader type-converter round-trips for recursive LIST.
- **Ingestion**: recursive list-like CSV parsing, recursive JSON LIST parsing (nested JSON nulls normalize to the declared child default), COPY into vertex and non-bundled edge tables.
- **Queries**: explicit `CAST` between LIST and ARRAY with recursive per-level ARRAY length validation (implicit conversion remains unsupported), execution-side LIST validity (NULL vs empty list), result-sink validity bitmap and uint32 offset bound.
- **Semantics**: top-level NULL persists as `[]`; NULL children normalize to declared defaults; WAL reuses the existing recursive Value redo (no new opcode or proto field); LIST primary keys / indexes / ORDER BY are out of scope for this round.

Verified with a full build plus targeted suites: column lifecycle/resize/type-contract tests, JSON/YAML/PB round-trip tests, TP snapshot-isolation/abort/WAL-replay tests covering LIST on both vertices and edges, and Python end-to-end tests (AP create/set/merge + close-reopen, COPY from CSV/JSON including negative cases, explicit CAST contract).

## Related issue number

Fixes #159
Part of #446

### CODE REVIEW

File: src/utils/property/list_property_column.cc [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+    auto offset = offsets_->get_view(row);
+    auto length = lengths_->get_view(row);
+    compact_offsets.set_value(row, tail);
+    compact_lengths.set_value(row, length);
+    compact_elements->resize(tail + length);
+    for (size_t i = 0; i < length; ++i) {
+      compact_elements->set_any(tail + i, elements_->get_any(offset + i), true);
+    }
+    tail += length;
+  }
```
[the remark this answers is not in the document]
Comment (zhanglei1949) [reply src/utils/property/list_property_column.cc]: Fixed in 2da6fbc9. Precomputed total element count by summing lengths_ across all rows, then resize compact_elements once before the copy loop. Eliminates the O(n^2) memcpy from repeated resize calls.

File: src/utils/property/list_property_column.cc [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+    for (size_t i = 0; i < children.size(); ++i) {
+      write_child(old_offset + i, children[i], insert_safe);
+    }
+    return;
+  }
+
+  auto new_offset = elements_->size();
+  elements_->resize(new_offset + children.size());
+  for (size_t i = 0; i < children.size(); ++i) {
+    write_child(new_offset + i, children[i], true);
```
[the remark this answers is not in the document]
Comment (zhanglei1949) [reply src/utils/property/list_property_column.cc]: Fixed in 2da6fbc9. Added an insert_safe guard: when list length changes and insert_safe is false, now throws STORAGE_EXCEPTION instead of silently resizing elements_. This aligns with the ColumnBase::set_any contract.

File: doc/source/cypher_manual/ddl_clause.md [hunk 3]
```diff
@@ -17,6 +17,9 @@ The following table lists the recommended syntax for defining default values for
 | `TIMESTAMP`       | `prop TIMESTAMP DEFAULT TIMESTAMP('1970-01-01')` | `TIMESTAMP('1970-01-01')`    |
 | `INTERVAL`        | `prop INTERVAL DEFAULT INTERVAL('0 year 0 month 0 day')` | `INTERVAL('0 year 0 month 0 day')`         |
 | `ARRAY`           | `prop INT32[3] DEFAULT [1, 2, 3]` | child defaults repeated to the fixed length, for example `[0, 0, 0]` for `INT32[3]` |
+| `LIST`             | `prop INT64[] DEFAULT [1, 2, 3]`  | `[]` (empty list)                  |
+
+List types (`T[]`) are variable-length and can hold any number of elements, including zero. Array types (`T[N]`) have a fixed length `N` that must be a positive integer.
```
Comment (longbinlai) [inline doc/source/cypher_manual/ddl_clause.md]: 感觉不太需要这句描述

File: doc/source/cypher_manual/ddl_clause.md [hunk 4]
```diff
@@ -124,6 +127,54 @@ CREATE NODE TABLE Matrix(
...
+CREATE (p:Person {id: 1, tags: CAST(['a', 'b'], 'STRING[]')});
+
+-- Error: bare list literal is not accepted
+CREATE (p:Person {id: 2, tags: ['a', 'b']});
+```
+
+List properties are supported in CSV and JSON bulk loading via `COPY FROM`. List values in CSV use bracket syntax `[1, 2, 3]`, and nested lists nest the brackets:
+
+```
+id|tags
```
Comment (longbinlai) [inline doc/source/cypher_manual/ddl_clause.md:171]: Please confirm that what type will be automatically inferred for this kind of data `[1,2,3]` in schema-free COPY/LOAD (it is array or list) @zhanglei1949 @shirly121

File: tools/python_bind/tests/test_db_list.py [hunk 5]
```diff
@@ -14,44 +14,177 @@
...
 
 from neug.database import Database
 
 
-def test_return_single_list(tmp_path):
-    db_dir = tmp_path / "return_list"
-    shutil.rmtree(db_dir, ignore_errors=True)
-    db_dir.mkdir()
-    db = Database(db_path=str(db_dir), mode="w")
+def _nested_list(value):
```
Comment (longbinlai) [inline tools/python_bind/tests/test_db_list.py:23]: Can we also test bulk load list/array data from csv/json/parquet?

File: doc/source/cypher_manual/ddl_clause.md (hunk 3 above)
Comment (zhanglei1949) [reply doc/source/cypher_manual/ddl_clause.md]: 我建议还是保留，虽然用户可能也知道List和Array的区别，但是我们还是明确定义一下

File: doc/source/cypher_manual/ddl_clause.md (hunk 4 above)
Comment (zhanglei1949) [reply doc/source/cypher_manual/ddl_clause.md:171]: In current implementation，the literal will be parsed as array in schema-free scenario. Please confirm @shirly121

File: tools/python_bind/tests/test_db_list.py (hunk 5 above)
Comment (zhanglei1949) [reply tools/python_bind/tests/test_db_list.py:23]: ok

File: doc/source/cypher_manual/ddl_clause.md (hunk 3 above)
Comment (longbinlai) [reply doc/source/cypher_manual/ddl_clause.md]: 你后面具体介绍list 可以在那里加哈

Comment (zhanglei1949) [reply doc/source/cypher_manual/ddl_clause.md]: 挪到后面了

File: tools/python_bind/tests/test_db_list.py (hunk 5 above)
Comment (zhanglei1949) [reply tools/python_bind/tests/test_db_list.py:23]: copy from csv/json的测试加了，copy from parque的建议后续单独PR再加：https://github.com/alibaba/neug/issues/836

File: doc/source/cypher_manual/ddl_clause.md (hunk 4 above)
Comment (shirly121) [reply doc/source/cypher_manual/ddl_clause.md:171]: > Please confirm that what type will be automatically inferred for this kind of data `[1,2,3]` in schema-free COPY/LOAD (it is array or list) @zhanglei1949 @shirly121

By default, the value is inferred as a string type. Implicit conversion to an array/list type is not supported. An explicit cast is required using CAST(val, 'INT32[]').

**Review by longbinlai — DISMISSED** [review]
LGTM

File: src/utils/io/read/json/json_reader.cc [hunk 6]
```diff
@@ -252,8 +276,13 @@ class JsonChunkSupplier : public IDataChunkSupplier {
               "Column '" + name +
               "' not found in JSON object in file: " + file_path_);
         }
-        builders[col]->push_back_elem(
-            parse_json_value(obj[name.c_str()], selected_types[col]));
+        const auto& json_value = obj[name.c_str()];
+        if (json_value.IsNull()) {
```
Comment (liulx20) [inline src/utils/io/read/json/json_reader.cc:280]: PR 把任意 JSON null 统一改为调用 builders[col]->push_back_null()，但 ContextArrayColumnBuilder 没有实现该接口，基类默认行为是 LOG(FATAL)。因此已有的 INT32[2] 等 ARRAY 字段在 JSON COPY 中出现 null 会导致加载进程退出，而不是按原有默认值/NULL 语义处理。应当仅对支持 nullable 的 builder 调用该方法，或为 ARRAY builder 实现正确的空值/默认值处理，并补回归测试

Comment (zhanglei1949) [reply src/utils/io/read/json/json_reader.cc:280]: https://github.com/alibaba/neug/pull/829/changes/6f7c3811471dfc726c7a4e637f3ce0d463f79bac

**Review by liulx20 — APPROVED** [review]
